src/spatialdata/_core/query/relational_query.py

```diff
@@ -129,6 +147,255 @@ def _filter_table_by_elements(
+                continue
+
+    return element_dict, None
+
+
+def _left_join_spatialelement_table(
+    element_dict: dict[str, dict[str, Any]], table: AnnData
+) -> tuple[dict[str, Any], AnnData]:
+    regions, region_column_name, instance_key = get_table_keys(table)
+    groups_df = table.obs.groupby(by=region_column_name)
+    joined_indices = None
+    for element_type, name_element in element_dict.items():
+        for name, element in name_element.items():
+            if name in regions:
+                group_df = groups_df.get_group(name)
+                table_instance_key_column = group_df[instance_key]  # This is always a series
+                if element_type in ["points", "shapes"]:
+                    element_indices = element.index
+                else:
+                    element_indices = _get_unique_label_values_as_index(element)
+
+                mask = table_instance_key_column.isin(element_indices)
+                if joined_indices is None:
+                    joined_indices = table_instance_key_column[mask].index
+                else:
+                    # in place append does not work with pd.Index
+                    joined_indices = joined_indices.append(table_instance_key_column[mask].index)
+            else:
+                warnings.warn(
+                    f"The element `{name}` is not annotated by the table. Skipping", UserWarning, stacklevel=2
+                )
+                continue
+
+    joined_table = table[joined_indices, :].copy() if joined_indices is not None else None
+
+    return element_dict, joined_table
+
+
+class JoinTypes(Enum):
+    """Available join types for matching elements to tables and vice versa."""
+
+    LEFT = left = partial(_left_join_spatialelement_table)
+    LEFT_EXCLUSIVE = left_exclusive = partial(_left_exclusive_join_spatialelement_table)
+    INNER = inner = partial(_inner_join_spatialelement_table)
+    RIGHT = right = partial(_right_join_spatialelement_table)
+    RIGHT_EXCLUSIVE = right_exclusive = partial(_right_exclusive_join_spatialelement_table)
+
+    def __call__(self, *args: Any) -> tuple[dict[str, Any], AnnData]:
+        return self.value(*args)
+
+    @classmethod
+    def get(cls, key: str) -> Callable[[dict[str, dict[str, Any]], AnnData], tuple[dict[str, Any], AnnData]] | None:
+        if key in cls.__members__:
+            return cls.__members__.get(key)
+        return None
+
+
+def join_sdata_spatialelement_table(
+    sdata: SpatialData, spatial_element_name: str | list[str], table_name: str, how: str = "LEFT"
+) -> tuple[dict[str, Any], AnnData]:
+    assert sdata.tables.get(table_name), f"No table with `{table_name}` exists in the SpatialData object."
+    table = sdata.tables[table_name]
+    if isinstance(spatial_element_name, str):
+        spatial_element_name = [spatial_element_name]
+
+    elements_dict: dict[str, dict[str, Any]] = {}
+    for name in spatial_element_name:
+        element_type, _, element = sdata._find_element(name)
```

Comment:
I'd replace with the public function `sdata[name]`

Reply:
Would cause duplication of throwing key error that is already present in `_find_element`. Unless we create `__get__`.

Reply:
True, I'd go with the `get`, or momentarily use again `_find_element()`.

Reply:
on second thought, can't do that as it is useful in the join functions to have the `element_type` because the behaviour is different for different elements.